### PR TITLE
Seed _get_train_sampler's generator with arg seed to improve reproducibility

### PR DESCRIPTION
**Pull request by dlwh**
... and make the world_size<=1 code path more similar to the others

# What does this PR do?

Seed the generator that is (sometimes) used in _get_train_sampler with self.args.seed to improve reproducibility/control over randomness.

https://github.com/huggingface/transformers/pull/11582/files introduced using torch generators (when possible) to improve reproducibility in trainer's _get_train_sampler. However, the seed in TrainingArguments isn't used explicitly. This is usually fine, but if there's any non-determinism in the number of calls to random generators in, e.g., model_init, this doesn't do what you want. (For instance, let's say you wanted to play with differently sized models but leave all else equal including data loading order.) 

The other branches of this method appear (I think) to all use this same seed, so that helps bring this in line too, but that's probably not important.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
No, hope that's ok.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
internal bugfix

- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

Suggesting @sgugger (cc @siddk)

### CODE REVIEW

**Comment by sgugger** [comment]
No. As mentioned in the other issues, this code is intended to be this way. The torch RNG has been seeded so this is completely deterministic.

**Comment by sgugger** [comment]
Oh, I read too fast and didn't catch your whole point. Thanks for expanding :-)
How about we turn this on and off with a flag? Since the `seed` is set by default, I'm afraid this will hinder users launching the train method several times (they'll get the same shuffle then, instead of different ones).

**Comment by dlwh** [comment]
That makes sense to me! How about:

* we add a new flag called "sampler_seed" (or "data_seed" or whatever you
want) which defaults to None, and
* we use the old behavior if it's None and otherwise use sampler_seed for
this generator and for all the other seeds in that method otherwise.

**Comment by sgugger** [comment]
Sounds good to me!

**Comment by dlwh** [comment]
not sure what's going on with the doc build. it looks unrelated to me, but please let know if I'm wrong.

Otherwise, I made data_seed along the lines discussed, so I think it's RFAL.

File: src/transformers/training_args.py [hunk 1]
```diff
@@ -220,6 +220,10 @@ class TrainingArguments:
         seed (`int`, *optional*, defaults to 42):
             Random seed that will be set at the beginning of training. To ensure reproducibility across runs, use the
             [`~Trainer.model_init`] function to instantiate the model if it has some randomly initialized parameters.
+        data_seed (`int`, *optional*, defaults to None):
```
Comment (sgugger) [inline src/transformers/training_args.py:223]:
```suggestion
        data_seed (`int`, *optional*):
```
Default to None is implied by optional.

File: src/transformers/training_args.py [hunk 2]
```diff
@@ -539,6 +543,7 @@ class TrainingArguments:
     )
     no_cuda: bool = field(default=False, metadata={"help": "Do not use CUDA even when it is available"})
     seed: int = field(default=42, metadata={"help": "Random seed that will be set at the beginning of training."})
+    data_seed: Optional[int] = field(default=None, metadata={"help": "Random seed to be used with data samplers."})
```
Comment (sgugger) [inline src/transformers/training_args.py:546]:
```suggestion
    data_seed: int = field(default=None, metadata={"help": "Random seed to be used with data samplers."})
```
Here we don't need the real type hint as it won't be seen in the doc.

**Review by sgugger — APPROVED** [review]
Thanks for amending your PR! Just left a few styling nits, but it's almost ready to merge otherwise :-)

**Comment by sgugger** [comment]
Thanks again for your PR!

**Pull request merged**